txpipe/twopoint_null_tests.py

```diff
@@ -766,3 +767,272 @@ def write_output(self, source_list, lens_list, meta, results):
+                    w_yp  = nmt.NmtWorkspace()
+                    w_yp.compute_coupling_matrix(field1, field2, b)
+    
+                    cl_coupled   = nmt.compute_coupled_cell(field1,field2)
+                    cl_decoupled = w_yp.decouple_cell(cl_coupled)
+                    ret[zj,zi,:]= cl_decoupled[3]
+                    tmp = np.concatenate([tmp,cl_decoupled[3]])
+
+            
+            # Compute covariance in a somewhat adhoc way by shuffling the pixel values.
+            covarr = np.zeros((len(tmp),self.config['Nsims']))
+
+            for k in tqdm(range(self.config['Nsims'])):
+                fields = {}
+                for i in range(nbin_source):
+                    idx1 = np.where(g1_maps[i]!=0)[0]
+                    idx2 = np.random.permutation(idx1)
+                    g1_maps[i][idx1]=g1_maps[i][idx2]
+                    g2_maps[i][idx1]=g2_maps[i][idx2]
+                    fields[i] = nmt.NmtField(mask, [g1_maps[i], g2_maps[i]], purify_e=False, purify_b=purify_b)
+
+                tmp = np.array([])
+                for zi in range(nbin_source):
+                    for zj in range(zi,nbin_source):
+                        
+                        field1 = fields[zi]
+                        field2 = fields[zj]
+                        w_yp  = nmt.NmtWorkspace()
+                        w_yp.compute_coupling_matrix(field1, field2, b)
+        
+                        cl_coupled   = nmt.compute_coupled_cell(field1,field2)
+                        cl_decoupled = w_yp.decouple_cell(cl_coupled)
+                        tmp = np.concatenate([tmp,cl_decoupled[3]])
+
+                covarr[:,k] = tmp
+             
+                    
+            self.ell     = b.get_effective_ells()
+            self.results = ret
+            self.cov     = np.cov(covarr)
+        
+
+        elif self.config["method"] == 'hybrideb':
+            '''
+            B-mode method of Becker and Rozo 2015 http://arxiv.org/abs/1412.3851
+            '''
+            print('--------------hybrideb-----------')
+
+            # Check if nbins is less than 1000, and throw warning.
+            if self.config['Ntheta']<1000:
+                print("WARNING: Calculating hybridEB using Ntheta<1000, which may lead to inaccurate results.")
+
+            # Computing the weights takes a few minutes so its a lot faster
+            # to precompute and load them in subsequent runs.
+            file_precomputed_weights = self.get_output("twopoint_data_fourier_pureB")+ f".precomputedweights.pkl"
+
+            if os.path.exists(file_precomputed_weights):
+                print(f"{self.rank} WARNING: Using precomputed weights from revious run: {file_precomputed_weights}")
+                with open(file_precomputed_weights, "rb") as f:
+                    geb_dict = pickle.load(f)
+            
+            else:
+                heb = hybrideb.HybridEB(self.config['theta_min'], self.config['theta_max'], self.config['Ntheta'])
```

Comment:
Can you add more comments throughout this section? The main bit where it does the loop through pairs is okay but some more explanation of the various things we need first would be useful.

Reply:
Added the following comments:

> 

            else:
            # To run this method we need to first compute the Fourier and real-space windows
            # hybrideb.HybridEB: Sets up the estimator by creating top hat filters in real-space
            # hybrideb.BinEB   : Calculates the intermediate filter quantities, namely fa, fb, M+, M- [Eqs 6-11]
            # hybrideb.GaussEB : Creates Gaussian windows in Fourier space
            heb = hybrideb.HybridEB(self.config['theta_min'], self.config['theta_max'], Ntheta)
            beb = hybrideb.BinEB(self.config['theta_min'], self.config['theta_max'], Ntheta)
            geb = hybrideb.GaussEB(beb, heb)
            
            # geb is in a fromat that can not be naturally saved so we convert it to a regular dictionary.
            geb_dict = {}
            for i in range(0,Nell):
                geb_dict[f"{i+1}"] = {}
                for j in range(0,6):
                    geb_dict[f"{i+1}"][f"{j+1}"]=geb(i)[j]
            
            # geb_dict stores the following information in a regular dictionary format
            # 1: theta in radians
            # 2: Fp  plus component of the real-space window function
            # 3: Fm  minus component of the real-space window function
            # 4: ell
            # 5: Wp  plus component of Fourier-space window function
            # 6: Wm  minus component of Fourier-space window function

            # 2 and 3 are used to get Xp/Xm      = sum((Fp*xip_hat +/- Fm*xim_hat)/2)    [eq1]
            # 5 and 6 are used to get <Xp>/<Xm>  = \int ell factors(ell) (Wp*Pe + Wm*Pb) [eq5]

            # Save this for subsequent runs since the windows don't change.